streams/src/test/java/org/apache/kafka/streams/tests/StreamsRepeatingIntegerKeyProducer.java

```diff
@@ -0,0 +1,108 @@
+ * can be provided in the config map with key of 'topics' and ';' delimited list of output topics
+ */
+public class StreamsRepeatingIntegerKeyProducer {
+
+    private static volatile boolean keepProducing = true;
+    private static int messageCounter = 0;
+
+    public static void main(String[] args) {
+        System.out.println("StreamsTest instance started");
+
+        final String kafka = args.length > 0 ? args[0] : "localhost:9092";
+        final String configString = args.length > 2 ? args[2] : null;
+
+        Map<String, String> configs = SystemTestUtil.parseConfigs(configString);
+        System.out.println("Using provided configs " + configs);
+
+        final int numMessages = configs.containsKey("num_messages") ? Integer.parseInt(configs.get("num_messages")) : 1000;
+
+        final Properties producerProps = new Properties();
+        producerProps.put(ProducerConfig.CLIENT_ID_CONFIG, "StreamsRepeatingIntegerKeyProducer");
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        producerProps.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
+
+        String value = "testingValue";
+        Integer key = 0;
+
+        Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
+            @Override
+            public void run() {
+                keepProducing = false;
+            }
+        }));
+
+        final String[] topics = configs.get("topics").split(";");
+        final int totalMessagesToProduce = numMessages * topics.length;
+
+        try (final KafkaProducer<String, String> kafkaProducer = new KafkaProducer<>(producerProps)) {
+
+            while (keepProducing && messageCounter < totalMessagesToProduce) {
+                for (final String topic : topics) {
+                    final ProducerRecord<String, String> producerRecord = new ProducerRecord<>(topic, key.toString(), value + key);
+                    kafkaProducer.send(producerRecord, new Callback() {
+                        @Override
+                        public void onCompletion(final RecordMetadata metadata, final Exception exception) {
+                            if (exception != null) {
+                                exception.printStackTrace(System.err);
+                                System.err.flush();
+                                if (exception instanceof TimeoutException) {
+                                    try {
+                                        // message == org.apache.kafka.common.errors.TimeoutException: Expiring 4 record(s) for data-0: 30004 ms has passed since last attempt plus backoff time
+                                        final int expired = Integer.parseInt(exception.getMessage().split(" ")[2]);
+                                        messageCounter -= expired;
```

Comment:
Should messageCounter be `volatile`? Not sure if this callback happens in the same or a different thread? \cc @guozhangwang

Reply:
Ack, adding as I think callback happens on different thread